### PR TITLE
parse X-Kifi-Client header

### PR DESCRIPTION
**Pull request by yingjieMiao**
@2is10 

cc: @thassss @jaredpetker

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala [hunk 1]
```diff
@@ -148,15 +174,13 @@ class HeimdalContextBuilder {
...
     this += ("userAgent", userAgent)
     userAgent match {
+      // These two are here for backwards compatiblity. Remove it soon!
       case UserAgent.iosAppRe(appName, appVersion, buildSuffix, device, os, osVersion) =>
         this += ("device", device)
         this += ("os", os)
         this += ("osVersion", osVersion)
         this += ("client", "Kifi App")
-        this += ("clientVersion", appVersion)
-        this += ("clientBuild", appVersion + buildSuffix)
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala:157]: If this whole section is here for backwards compatibility, then we probably don’t want to remove any of it yet.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala [hunk 2]
```diff
@@ -166,9 +190,7 @@ class HeimdalContextBuilder {
         this += ("device", agent.getDeviceCategory.getName)
         this += ("os", agent.getOperatingSystem.getFamilyName)
         this += ("osVersion", agent.getOperatingSystem.getName)
-        this += ("client", agent.getName)
-        this += ("clientVersion", agent.getVersionNumber.getMajor)
-        this += ("clientBuild", agent.getVersionNumber.toVersionString)
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala:171]: We may want to leave these three lines here for a few more weeks, until everyone using the browser extension has updated to a version that sends the X-Kifi-Client header.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala (hunk 1 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala:157]: Ah, good catch. Will put them back.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala (hunk 2 above)
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala:171]: Do `clientVersion` and `clientBuild` currently contain the browser version and build number? If so, that’s very useful and we’ll probably want to continue sending them forever, and we may need to choose different property names for the browser extension version.

Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala:171]: good suggestion. It does contain browser version and build number. I will choose different names for them.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala [hunk 3]
```diff
@@ -122,10 +122,36 @@ class HeimdalContextBuilder {
...
+    }
+
+    val parts = clientVersion.split("\\s")
+    if (parts.size == 2) {
+      val (client, version) = (parts(0), parts(1))
+      client match {
+        case "BE" => addKifiClientVersion(agent.getName, KifiExtVersion(version)) // Chrome, Firefox, Chromium, Opera, etc
+        case "iOS" => addKifiClientVersion("Kifi App", KifiIPhoneVersion(version))
+        case "iOS Extension" => addKifiClientVersion("Kifi iOS Extension", KifiIPhoneVersion(version))
+        case "Android" => addKifiClientVersion("android", KifiAndroidVersion(version))
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala]: For `"client"`, I think we just decided with Jen to use `"browserExtension"`, `"iOsApp"`,`"iOsExtension"`, `"androidApp"`, and `"androidExtension"`.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala [hunk 4]
```diff
@@ -166,9 +192,9 @@ class HeimdalContextBuilder {
         this += ("device", agent.getDeviceCategory.getName)
         this += ("os", agent.getOperatingSystem.getFamilyName)
         this += ("osVersion", agent.getOperatingSystem.getName)
-        this += ("client", agent.getName)
-        this += ("clientVersion", agent.getVersionNumber.getMajor)
-        this += ("clientBuild", agent.getVersionNumber.toVersionString)
+        this += ("agent", agent.getName)
+        this += ("agentVersion", agent.getVersionNumber.getMajor)
+        this += ("agentBuild", agent.getVersionNumber.toVersionString)
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala]: Jen says we don’t need `"agentBuild"`.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala [hunk 5]
```diff
@@ -122,10 +122,36 @@ class HeimdalContextBuilder {
...
   }
 
+  private def addVersion(request: RequestHeader): Unit = {
+
+    val clientVersion = request.headers.get("X-Kifi-Client") getOrElse ""
+    val userAgentStr = request.headers.get("User-Agent").getOrElse("")
+    val agent = UserAgent.parser.parse(userAgentStr)
+
+    def addKifiClientVersion(clientName: String, version: KifiVersion): Unit = {
+      this += ("clientVersion", version.major + "." + version.minor + "." + version.patch)
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala]: `version.minor` and `version.patch` can be `-1`, which signifies that they should be omitted. I’ll add a `toStringWithoutBuild` method to `KifiVersion` for you to use here.

Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala]: Thanks!

**Comment by yingjieMiao** [comment]
@2is10 

Updates checklist:
1. client field: browserExtention, iOSApp, iOSExtention, androidApp, (androidExtention not added yet)
2. do not send agentBuild
3. use toStringWithoutBuild
4. Stop using 'Kifi App', 'android'

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala [hunk 6]
```diff
@@ -122,10 +122,34 @@ class HeimdalContextBuilder {
...
+
+    val parts = clientVersion.trim.split("\\s")
+    if (parts.size == 2) {
+      val (client, version) = (parts(0), parts(1))
+      client match {
+        case "BE" => add("browserExtension", KifiExtVersion(version))
+        case "iOS" => add("iOSApp", KifiIPhoneVersion(version))
+        case "iOS Extension" => add("iOSExtension", KifiIPhoneVersion(version))
+        case "Android" => add("androidApp", KifiAndroidVersion(version))
+        case _ =>
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala]: worth logging or airbraking here?

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala [hunk 7]
```diff
@@ -148,27 +172,26 @@ class HeimdalContextBuilder {
   def addUserAgent(userAgent: String): Unit = {
     this += ("userAgent", userAgent)
     userAgent match {
+      // These two are here for backwards compatiblity. Remove it soon!
       case UserAgent.iosAppRe(appName, appVersion, buildSuffix, device, os, osVersion) =>
         this += ("device", device)
         this += ("os", os)
         this += ("osVersion", osVersion)
-        this += ("client", "Kifi App")
+        this += ("client", "iOSApp")
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala:181]: I’m not sure it’s a good idea to change values in code paths that exist solely for backwards compatibility.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala (hunk 6 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala]: I will add a log first.

**Comment by 2is10** [comment]
Nice work. Feel free to merge when you’re happy with it.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala (hunk 7 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalContext.scala:181]: I think this makes the tracking more consistent. We, especially product 
people, want to discard 'Kifi App' asap.

**Comment by yingjieMiao** [comment]
merging. thanks for the help!

**Pull request merged**